src/ILOPoolSale.sol

```diff
@@ -0,0 +1,217 @@
+        require(SALE_END > block.timestamp, 'ES');
+        _;
+    }
+
+    function initialize(InitParams calldata params) external override {
+        _validateVestSchedule(params.vestingSchedule);
+        _initialize(params.baseParams);
+
+        for (uint256 i = 0; i < params.vestingSchedule.length; i++) {
+            _vestingSchedule.push(params.vestingSchedule[i]);
+        }
+
+        (SALE_START, SALE_END, MIN_RAISE, MAX_RAISE) = (
+            params.saleParams.start,
+            params.saleParams.end,
+            params.saleParams.minRaise,
+            params.saleParams.maxRaise
+        );
+
+        emit ILOPoolSaleInitialized(
+            params.baseParams,
+            params.saleParams,
+            params.vestingSchedule
+        );
+    }
+
+    function buy(
+        uint256 raiseAmount,
+        address recipient
+    )
+        external
+        override
+        duringSale
+        whenNotCancelled
+        nonReentrant
+        returns (uint256 tokenId)
+    {
+        require(raiseAmount > 0, 'ZA');
+
+        // transfer fund into contract
+        TransferHelper.safeTransferFrom(
+            PAIR_TOKEN,
+            msg.sender,
+            address(this),
+            raiseAmount
+        );
+
+        // check if raise amount over capacity
+        require(MAX_RAISE - TOTAL_RAISED >= raiseAmount, 'MR');
+        TOTAL_RAISED += raiseAmount;
+
+        // if investor already have a position, just increase raise amount and liquidity
+        // otherwise, mint new nft for investor and assign vesting schedules
+        if (balanceOf(recipient) == 0) {
+            _mint(recipient, (tokenId = _nextId++));
+            _positionVests[tokenId].schedule = _vestingSchedule;
+        } else {
+            tokenId = tokenOfOwnerByIndex(recipient, 0);
+        }
+
+        Position storage _position = _positions[tokenId];
+
+        uint256 _allocation = allocation(recipient);
+        // we need to check `_position.raiseAmount < _allocation`
+        // because project admin can change whitelist during sale.
+        // so, technicaly, `_position.raiseAmount` can bigger than `_allocation`
+        // second reason is to avoid overflow of subtraction.
+        require(
+            _position.raiseAmount < _allocation &&
+                raiseAmount <= _allocation - _position.raiseAmount,
+            'UC'
+        );
+        _position.raiseAmount += raiseAmount;
+
+        emit Buy(recipient, tokenId, raiseAmount);
+    }
+
+    /// @inheritdoc IILOPoolBase
+    function launch(
+        address uniV3PoolAddress,
+        PoolAddress.PoolKey calldata poolKey,
+        uint160 sqrtPriceX96
+    ) external override onlyManager onlyInitializedProject whenNotCancelled {
+        if (block.timestamp < SALE_END) {
+            revert('ST');
+        }
+        if (TOTAL_RAISED < MIN_RAISE) {
+            revert('MR');
+        }
+        uint128 liquidity = _launchLiquidity(
```

Comment:
Missing lines to transfer the raised amount back to the project admin.

The transfer should happen before the `_launchLiquidity`, to make sure project has enough money for launching liquidity.